### PR TITLE
Adding more async examples (with AMPHP)

### PR DESCRIPTION
**Pull request by kinimodmeyer**
- Mentioning in the documentation also the existing AMPHP adapter `AmpPromiseAdapter`
- Adding two more async examples for amphp.
   - One which is similar to the existing reactphp example (http server)
   - Another one which can be run in environments with php fpm/apache
- Group async examples into a parent 04-async-php folder
- Remove icicle.io from the list (is long ago deprecated in favor of Amp v2.0)
- Remove notice regarding version 0.9.0. seems long outdated as we are at version v15

In general AMPHP (& ReactPHP) reworking their frameworks to work natively with PHP8.1 fibers.
With new AMPv3 we might not need the "call" anymore. I´m unsure if the adaper might need a change too.
In a future/next iteration i will test/migrate this.

### CODE REVIEW

File: examples/04-async-php/amphp/event-loop/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## Event loop with AMPHP
+
+While you can create an own http-server with PHP GraphQL & AMPHP (see "http-server" example)
```
Comment (spawnia) [inline examples/04-async-php/amphp/event-loop/README.md:3]:
```suggestion
While you can create your own http-server with PHP GraphQL & AMPHP (see "http-server" example),
```

File: examples/04-async-php/amphp/event-loop/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
...
+this example aims to give you an event-loop on traditional request model (apache, fpm, ...)
```
Comment (spawnia) [inline examples/04-async-php/amphp/event-loop/README.md:4]:
```suggestion
this example aims to give you an event-loop combined with the traditional request model (apache, fpm, ...).
```

File: examples/04-async-php/amphp/event-loop/graphql.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+require_once __DIR__ . '/../../../../vendor/autoload.php';
+
+use GraphQL\Executor\ExecutionResult;
+use GraphQL\Executor\Promise\Adapter\AmpPromiseAdapter;
+use GraphQL\GraphQL;
+
+$schema = require_once __DIR__ . '/../schema.php';
+
+\Amp\Loop::run(function () use ($schema) {
```
Comment (spawnia) [inline examples/04-async-php/amphp/event-loop/graphql.php:14]: import

Comment (spawnia) [inline examples/04-async-php/amphp/event-loop/graphql.php:14]: return type?

File: examples/04-async-php/amphp/event-loop/graphql.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        new AmpPromiseAdapter(),
+        $schema,
+        $input['query'],
+        null,
+        null,
+        $input['variables'] ?? null,
+        $input['operationName'] ?? null
+    );
+    $resultArray = [];
+    $promise->then(function(ExecutionResult $result) use (&$resultArray) {
```
Comment (spawnia) [inline examples/04-async-php/amphp/event-loop/graphql.php:26]: void

File: examples/04-async-php/amphp/event-loop/graphql.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        $resultArray = $result->toArray();
+    });
+    echo json_encode($resultArray);
```
Comment (spawnia) [inline examples/04-async-php/amphp/event-loop/graphql.php:29]: why not echo in `then()`?

File: examples/04-async-php/amphp/http-server/graphql.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+use Amp\Http\Server\Request;
+use Amp\Http\Server\RequestHandler\CallableRequestHandler;
+use Amp\Http\Server\Response;
+use GraphQL\Executor\ExecutionResult;
+use GraphQL\Executor\Promise\Adapter\AmpPromiseAdapter;
+use GraphQL\GraphQL;
+
+$schema = require_once __DIR__ . '/../schema.php';
+
+\Amp\Loop::run(function () use ($schema) {
```
Comment (spawnia) [inline examples/04-async-php/amphp/http-server/graphql.php:18]: import + return type

File: examples/04-async-php/amphp/http-server/graphql.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    $sockets = [
+        Amp\Socket\Server::listen("localhost:8080"),
```
Comment (spawnia) [inline examples/04-async-php/amphp/http-server/graphql.php:20]: import

File: examples/04-async-php/amphp/http-server/graphql.php [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    ];
+
+    $server = new HttpServer($sockets, new CallableRequestHandler(function (Request $request) use ($schema) {
```
Comment (spawnia) [inline examples/04-async-php/amphp/http-server/graphql.php:23]: retuurn type?

File: examples/04-async-php/amphp/http-server/graphql.php [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        $input = json_decode(yield $request->getBody()->buffer(), true);
+        $query = $input['query'];
+        $variableValues = $input['variables'] ?? null;
+        $promise = GraphQL::promiseToExecute(new AmpPromiseAdapter(), $schema, $query, [], null, $variableValues);
+        $promise = $promise->then(function(ExecutionResult $result) {
+            return new Response(
```
Comment (spawnia) [inline examples/04-async-php/amphp/http-server/graphql.php:29]: short closure + return type

File: examples/04-async-php/amphp/schema.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
+<?php declare(strict_types=1);
+
+/*
+ * The article/product resolver get now resolved asynchronous.
+ * This can have big advantages if you fetch data from different microservices.
+ * Keep in mind, everything in "call" should now be non-blocking, checkout out async libraries like (amphp/mysql, amphp/http-client).
+ */
+return new GraphQL\Type\Schema([
+    'query' => new GraphQL\Type\Definition\ObjectType([
```
Comment (spawnia) [inline examples/04-async-php/amphp/schema.php:9]: use imports

File: examples/04-async-php/amphp/schema.php [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+        'name' => 'Query',
+        'fields' => [
+            'product' => [
+                'type' => GraphQL\Type\Definition\Type::string(),
+                'resolve' => function () {
+                    return \Amp\call(
+                        function() {
```
Comment (spawnia) [inline examples/04-async-php/amphp/schema.php:16]: short closure + type

**Review by spawnia — COMMENTED** [review]
Thanks for adding this, I like the direction

**Comment by kinimodmeyer** [comment]
@spawnia thanks for the fast review, i modified it regardless your valid suggestions.

Regarding the psalm errors, are you fine by adding "amp/http-server" to composer require-dev? (& suggest)
(should not make a lot dependencies as we have amp framework already)

Or do you want to modify psalm for the examples/* directories?
(then we **could** remove react/http|promise dependency too)

File: examples/04-async-php/amphp/http-server/graphql.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        $input = json_decode(yield $request->getBody()->buffer(), true);
+        $query = $input['query'];
+        $variableValues = $input['variables'] ?? null;
+        $promise = GraphQL::promiseToExecute(new AmpPromiseAdapter(), $schema, $query, [], null, $variableValues);
+        $promise = $promise->then(function(ExecutionResult $result): Response {
+            $data = json_encode($result->toArray());
+            return new Response(
+                200,
+                ['Content-Type' => 'application/json'],
+                $data !== false ? $data : null
```
Comment (spawnia) [inline examples/04-async-php/amphp/http-server/graphql.php:35]: Use `JSON_THROW_ON_ERROR` instead (in every `json_encode()` example)

File: examples/04-async-php/amphp/http-server/graphql.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            );
+        });
+        return $promise->adoptedPromise;
+    }), new Psr\Log\NullLogger);
```
Comment (spawnia) [inline examples/04-async-php/amphp/http-server/graphql.php:39]: import

File: examples/04-async-php/amphp/schema.php [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+use function Amp\call;
+
+return new Schema([
+    'query' => new ObjectType([
+        'name' => 'Query',
+        'fields' => [
+            'product' => [
+                'type' => Type::string(),
+                'resolve' => function (): Promise {
+                    return call(
```
Comment (spawnia) [inline examples/04-async-php/amphp/schema.php:22]: can be a short closure

File: examples/04-async-php/amphp/schema.php [hunk 15]
```diff
@@ -0,0 +1,39 @@
+<?php declare(strict_types=1);
+
+/*
+ * The article/product resolver get now resolved asynchronous.
```
Comment (spawnia) [inline examples/04-async-php/amphp/schema.php:4]:
```suggestion
 * The resolvers for "product" and "article" resolver run asynchronously.
```

File: examples/04-async-php/amphp/schema.php [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+ * This can have big advantages if you fetch data from different microservices.
+ * Keep in mind, everything in "call" should now be non-blocking, checkout out async libraries like (amphp/mysql, amphp/http-client).
```
Comment (spawnia) [inline examples/04-async-php/amphp/schema.php:6]:
```suggestion
 * Keep in mind everything in "call" should be non-blocking, check out async libraries like amphp/mysql, amphp/http-client.
```

**Review by spawnia — CHANGES_REQUESTED** [review]

**Pull request merged**